File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/MusicGroupNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;

namespace SoundPlayground.Parser.AbstractSyntaxTree
{
    public class MusicGroupNode : MusicNode
    {
        public MusicNode Expression { get; set; }

        public MusicGroupNode ( MusicNode expression ) {
            Expression = expression;
        }

        public override IEnumerable<Note> GetCommands ( Context context ) {
            Context forked = context.Fork();

            try {
                foreach ( var note in Expression.GetCommands( forked ) ) {
                    yield return note;
                }
            } finally {
                context.Join( forked );
            }
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/music_group_node.py
from .music_node import MusicNode

class MusicGroupNode( MusicNode ):
    def __init__ ( self, expression, position : (int, int) = None ):
        super().__init__( position )
        
        self.expression = expression

    def get_events ( self, context ):
        forked = context.fork()

        try:
            for event in self.expression.eval( context ):
                yield event
        finally:
            context.join( forked )
